ingest/importer/importer.py

```diff
@@ -51,28 +55,33 @@ def dry_run_import_file(self, file_path, project_uuid=None):
 
         return entity_map, []
 
-    def import_file(self, file_path, submission_url, is_update, project_uuid=None) -> Tuple[Submission, TemplateManager]:
-        submission = None
+    def import_file(self, file_path, submission_url, is_update=False, project_uuid=None) -> Tuple[Submission, TemplateManager]:
         try:
-            spreadsheet_json, template_mgr, errors = self.generate_json(file_path, is_update, project_uuid)
-            if not errors:
-                entity_map = self._process_links_from_spreadsheet(template_mgr, spreadsheet_json)
+            submission = None
+            template_mgr = None
+            spreadsheet_json, template_mgr, errors = self.generate_json(file_path, is_update, project_uuid=project_uuid)
+            entity_map = EntityMap.load(spreadsheet_json)
 
-                # TODO the submission_url should be passed to the IngestSubmitter instead
-                submitter = IngestSubmitter(self.ingest_api)
-                submission = submitter.submit(entity_map, submission_url)
-                self.logger.info(f'Submission in {submission_url} is done!')
-            else:
+            if errors:
                 self.report_errors(submission_url, errors)
+            elif is_update:
+                self.submitter.update_entities(entity_map)
+            else:
+                entity_map = self._process_links_from_spreadsheet(template_mgr, spreadsheet_json)
+                submission = self.submitter.add_entities(entity_map, submission_url)
+                return submission, template_mgr
         except HTTPError as httpError:
             status = httpError.response.status_code
             text = httpError.response.text
             importer_error = ImporterError(f'Received an HTTP {status} from  {httpError.request.url}: {text}')
             self.ingest_api.create_submission_error(submission_url, importer_error.getJSON())
+            return None, template_mgr
         except Exception as e:
             self.ingest_api.create_submission_error(submission_url, ImporterError(str(e)).getJSON())
             self.logger.error(str(e), exc_info=True)
-        else:
+            return None, template_mgr
+        finally:
```

Comment:
Ah, I just learned now, it will always execute even if there are returns in the except block. So i think, it's not bad to retain it so that you know in the logs that it's finished.